src/character.cpp

```diff
@@ -813,92 +813,175 @@ void Character::add_msg_player_or_say( const game_message_params &params,
     Messages::add_msg( params, player_msg );
 }
 
-int Character::effective_dispersion( int dispersion ) const
+int Character::effective_dispersion( int dispersion, bool zoom ) const
+{
+    return get_character_parallax( zoom ) + dispersion;
+}
+
+int Character::get_character_parallax( bool zoom ) const
 {
     /** @EFFECT_PER penalizes sight dispersion when low. */
-    dispersion += ranged_per_mod();
+    int character_parallax = zoom ? static_cast<int>( ranged_per_mod() * 0.25 ) : ranged_per_mod();
+    character_parallax += get_modifier( character_modifier_ranged_dispersion_vision_mod );
 
-    dispersion += get_modifier( character_modifier_ranged_dispersion_vision_mod );
+    return std::max( static_cast<int>( std::round( character_parallax ) ), 0 );
+}
 
-    return std::max( static_cast<int>( std::round( dispersion ) ), 0 );
+static double modified_sight_speed( double aim_speed_modifier, double effective_sight_dispersion,
+                                    double recoil )
+{
+    if( recoil <= effective_sight_dispersion ) {
+        return 0;
+    }
+    // When recoil tends to effective_sight_dispersion, the aiming speed bonus will tend to 0
+    double attenuation_factor = 2 * logarithmic( -std::pow( effective_sight_dispersion /
+                                ( recoil - effective_sight_dispersion ), 0.5 ) );
```

Comment:
This seems needlessly obtuse, as recoil approaches effective_sight_dispersion, effective_sight_dispersion / ( recoil - effective_sight_dispersion ) becomes large, but then you take the square root to make it grow more slowly, but then exponentiate it to make it grow much faster, then take the reciprocal.
This is exactly the kind of thing we use logarithmic_range for as it maps the linear change to a nice decay curve without too many steps.
```suggestion
    double attenuation_factor = 2 * logarithmic_range( effective_light_dispersion, RECOIL_MAX, recoil );
```

Reply:
You are right, this code looks too verbose
At the beginning, I didn't realize that using this function can also make the aiming speed decay quickly when recoil reaches 2*effective_sight_dispersion instead of (MAX_RECOIL+effective_sight_dispersion)/2.

Reply:
Double check, but I think what I suggested will have the same overall shape and endpoints as your formula.

Reply:
emmm..Using logarithmic_range is a good idea, but the parameter should not be like this.
`    double attenuation_factor = 1 - logarithmic_range( effective_sight_dispersion,
                                3 * effective_sight_dispersion, recoil );`

Reply:
no problem, I haven't had a chance to actually build and test the output, so I'm not surprised if I had some details wrong.